python_modules/dagster/dagster/_core/execution/context/system.py

```diff
@@ -806,7 +806,13 @@ def step_materializes_assets(self) -> bool:
     def input_asset_records(self) -> Optional[Mapping[AssetKey, Optional["EventLogRecord"]]]:
         return self._input_asset_records
 
-    def fetch_input_asset_records(self) -> None:
+    def get_input_asset_record(self, key: AssetKey) -> Optional["EventLogRecord"]:
+        if not key in self._input_asset_records:
+            self._fetch_input_asset_record(key)
+        return self._input_asset_records[key]
+
+    # "external" refers to records for inputs generated outside of this step
+    def fetch_external_input_asset_records(self) -> None:
```

Comment:
I think the logic here is still slightly off, as it assumes that any asset upstream of an asset produced by this step is "external".

Take for example a multi-asset that has the following internal dependency structure:

```
    a
  /   \
 b     c
```

and let's imagine "`a`" depends on "`a'`" (produced by a different step).

When we run this function to materialize a, b, and c, upstream_assets_for_asset("b")/upstream_assets_for_asset("c") will both return AssetKey("a"). This means that self._input_asset_records[AssetKey("a")] will be populated when this function is called, and so it'll have an old version, rather than the version created when running this step.

however, if we run this function to materialize just b and c, then the behavior described above is actually correct, as you don't expect to get a new version of a, so you should query for whatever the old version was.

the asset_info returned by asset_info_for_output has an "is_required" property, which will tell you if, for the current execution, that asset key is required to be executed. (generally, if is_required is `False`, then we do not expect it to be generated, although we'll just emit a warning rather than an error if it is.

I think this information should be used to get a set of required_output_keys (rather than all output keys), but mostly I'm just advocating for tests of these sorts of scenarios